py-polars/tests/unit/operations/aggregation/test_aggregations.py

```diff
@@ -27,6 +27,276 @@ def test_quantile_expr_input() -> None:
     )
 
 
+def test_quantile_varying_by_group_20951() -> None:
+    df = pl.DataFrame(
+        {
+            "value": [1, 2, 1, 2],
+            "quantile": [0, 0, 1, 1],
+        }
+    )
+    result = (
+        df.group_by(pl.col.quantile)
+        .agg(pl.col.value.quantile(pl.col.quantile.first()))
+        .sort("quantile")
+    )
+
+    expected = pl.DataFrame(
+        {
+            "quantile": [0, 1],
+            "value": [1.0, 2.0],
+        }
+    )
+    assert_frame_equal(result, expected)
+
+    df = pl.DataFrame(
+        {
+            "group": ["a", "a", "a", "b", "b", "b"],
+            "value": [1, 2, 3, 10, 20, 30],
+            "q": [0.0, 0.0, 0.0, 0.5, 0.5, 0.5],
+        }
+    )
+    result = (
+        df.group_by("group").agg(pl.col.value.quantile(pl.col.q.first())).sort("group")
+    )
+
+    expected = pl.DataFrame(
+        {
+            "group": ["a", "b"],
+            "value": [1.0, 20.0],
+        }
+    )
+    assert_frame_equal(result, expected)
+
+
+@pytest.mark.parametrize(
+    "method",
+    ["nearest", "higher", "lower", "midpoint", "linear", "equiprobable"],
+)
+def test_quantile_varying_by_group_methods(method: QuantileMethod) -> None:
+    df = pl.DataFrame(
+        {
+            "group": ["a", "a", "a", "a", "b", "b", "b", "b"],
+            "value": [1, 2, 3, 4, 10, 20, 30, 40],
+            "q": [0.25] * 4 + [0.75] * 4,
+        }
+    )
+    result = (
+        df.group_by("group")
+        .agg(pl.col.value.quantile(pl.col.q.first(), interpolation=method))
+        .sort("group")
+    )
+    assert result.shape == (2, 2)
+    assert result["group"].to_list() == ["a", "b"]
+
+
+def test_quantile_varying_by_group_float32() -> None:
```

Comment:
I feel like these are way too many individual tests. You can search around for `pytest.mark.parametrize` to see how to parametrize the tests. You can verify against a naive Python implementation which groups into lists then manually calls `quantile` on each list with a specific quantile.